src/dev/impl/DevToys/ViewModels/Tools/Graphic/ImageConverter/ImageConverterToolViewModel.cs

```diff
@@ -0,0 +1,249 @@
+                {
+                    return;
+                }
+
+                IReadOnlyList<IStorageItem>? storageItems = await parameters.DataView.GetStorageItemsAsync();
+                if (storageItems is null || storageItems.Count == 0)
+                {
+                    return;
+                }
+
+                foreach (IStorageItem storageItem in storageItems)
+                {
+                    if (storageItem is StorageFile file)
+                    {
+                        string fileExtension = Path.GetExtension(file.Name);
+
+                        if (SupportedFileExtensions.Any(ext => string.Equals(ext, fileExtension, StringComparison.OrdinalIgnoreCase)))
+                        {
+                            QueueNewConversion(file);
+                        }
+                        else
+                        {
+                            HasInvalidFilesSelected = true;
+                        }
+                    }
+                    else
+                    {
+                        HasInvalidFilesSelected = true;
+                    }
+                }
+            }).ConfigureAwait(false);
+        }
+
+        #endregion
+
+        #region SelectFilesBrowseCommand
+
+        public IAsyncRelayCommand SelectFilesBrowseCommand { get; }
+
+        private async Task ExecuteSelectFilesBrowseCommandAsync()
+        {
+            await ThreadHelper.RunOnUIThreadAsync(async () =>
+            {
+                HasInvalidFilesSelected = false;
+
+                var filePicker = new FileOpenPicker
+                {
+                    ViewMode = PickerViewMode.List,
+                    SuggestedStartLocation = PickerLocationId.ComputerFolder
+                };
+
+                for (int i = 0; i < SupportedFileExtensions.Length; i++)
+                {
+                    filePicker.FileTypeFilter.Add(SupportedFileExtensions[i]);
+                }
+
+                IReadOnlyList<StorageFile> files = await filePicker.PickMultipleFilesAsync();
+                for (int i = 0; i < files.Count; i++)
+                {
+                    QueueNewConversion(files[i]);
+                }
+            });
+        }
+
+        #endregion
+
+        #region DeleteAllCommand
+
+        public IRelayCommand DeleteAllCommand { get; }
+
+        private void ExecuteDeleteAllCommand()
+        {
+            ConversionWorkQueue.Clear();
+        }
+
+        #endregion
+
+        #region SaveAllCommand
+
+        public IAsyncRelayCommand SaveAllCommand { get; }
+
+        private async Task ExecuteSaveAllCommandAsync()
+        {
+            var works = ConversionWorkQueue.ToList();
+
+            var folderPicker = new FolderPicker
+            {
+                ViewMode = PickerViewMode.List
+            };
+
+            StorageFolder? selectedFolder = await folderPicker.PickSingleFolderAsync();
+
+            if (selectedFolder is not null)
+            {
```

Comment:
If I'm not mistaking, even if this method is a Task, we're still running on the UI Thread (you can verify it with `ThreadHelper.ThrowIfOnUIThread`).
This is fine until line 210. We need to be on the UI Thread in order to call PickSingleFolderAsync.

However, since the loop below can potentially treat hundreds files, I'd heavily suggest to run this part off the UI thread, perhaps showing a progress bar if needed (not sure if it really make sense but we can brainstorm it).

Reply:
I think you're right, process a lot of files in the UI thread may be a hard job. Instead of that, I have applied another approach: after ```PickSingleFolderAsync()``` method as many tasks as images to be converted will be created, and then those tasks will be executed asynchronusly using ```Task.WhenAll``` [so we won´t be tie up a thread to just hanging](https://stackoverflow.com/questions/6123406/waitall-vs-whenall).

All tasks created [will run on a ThreadPool thread](https://docs.microsoft.com/en-us/dotnet/api/system.threading.threadpool?view=net-6.0#remarks) so we will use background threads for that.

It is worth highlighting that the user will see how ImageConversionWorkItems representation in the UI will be gradually phased out (due to ```work.DeleteCommand.Execute(null)```), so we don´t need to show a progress bar or something like that to show the user we're doing any kind of background job.